libs/base/Data/List.idr

```diff
@@ -117,6 +119,27 @@ find : (p : a -> Bool) -> (xs : List a) -> Maybe a
 find p [] = Nothing
 find p (x::xs) = if p x then Just x else find p xs
 
+||| Find the index and proof of InBounds of the first element (if exists) of a
+||| list that satisfies the given test, else `Nothing`.
+public export
+findIndex : (a -> Bool) -> (xs : List a) -> Maybe (n : Nat ** InBounds n xs)
```

Comment:
Why not to use `Subset Nat (\n => InBounds n xs)` insead of `(n : Nat ** InBounds n xs)` in the return type, just to not to have an `InBounds` at runtime? As far as I remember, `InBounds` argument has always `0` quantity in functions that use it (like `index`), so there is no need to have it at runtime.

Another similar question, why not to go more radical and to have a `Fin (length xs)` instead of `(n : Nat ** InBounds n xs)` or `Subset Nat (\n => InBounds n xs)`. As far as we have a runtime optimisation for `Fin`s to be fast integers, it feels plausible. (Moreover, I personally have a finding function returning `Fin (length xs)` in my codebase and feel it better than `InBounds` version). However, `Fin`-based vs. `InBounds`-based approaches can be a holy war.

Reply:
I didn't mention `Subset` only because we don't have sugar for it, therefore I personally consider it painful to use...

Reply:
In case of `Fin (length xs)`, `Fin` is indexed by a computation. I don't have a feel for particular circumstances when it can turn out to be a nuisance. But I do know that sometimes in makes computations stuck which is to be avoided...